File: business/src/main/java/com/kevinguanchedarias/owgejava/business/mission/MissionConfigurationBo.java
package com.kevinguanchedarias.owgejava.business.mission;

import com.kevinguanchedarias.owgejava.business.ConfigurationBo;
import com.kevinguanchedarias.owgejava.entity.Configuration;
import com.kevinguanchedarias.owgejava.enumerations.MissionType;
import com.kevinguanchedarias.owgejava.exception.ProgrammingException;
import com.kevinguanchedarias.owgejava.exception.SgtBackendInvalidInputException;
import com.kevinguanchedarias.owgejava.repository.ConfigurationRepository;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @since 0.11.0
 */
@Service
@AllArgsConstructor
@Slf4j
public class MissionConfigurationBo {
    private final ConfigurationBo configurationBo;
    private final ConfigurationRepository configurationRepository;

    public static final String MISSION_TIME_EXPLORE_KEY = "MISSION_TIME_EXPLORE";
    public static final String MISSION_TIME_GATHER_KEY = "MISSION_TIME_GATHER";
    public static final String MISSION_TIME_ESTABLISH_BASE_KEY = "MISSION_TIME_ESTABLISH_BASE";
    public static final String MISSION_TIME_ATTACK_KEY = "MISSION_TIME_ATTACK";
    public static final String MISSION_TIME_CONQUEST_KEY = "MISSION_TIME_CONQUEST";
    public static final String MISSION_TIME_COUNTERATTACK_KEY = "MISSION_TIME_COUNTERATTACK";
    public static final String MISSION_TIME_DEPLOY_KEY = "MISSION_TIME_DEPLOY";

    public static final String DEFAULT_TIME_EXPLORE = "60";
    public static final String DEFAULT_TIME_GATHER = "900";
    public static final String DEFAULT_TIME_ESTABLISH_BASE = "43200";
    public static final String DEFAULT_TIME_ATTACK = "600";
    public static final String DEFAULT_TIME_CONQUEST = "86400";
    public static final String DEFAULT_TIME_COUNTERATTACK = "60";
    public static final String DEFAULT_TIME_DEPLOY = "60";

    private static class MissionTimeStore {
        private final String time;
        private final String description;

        public MissionTimeStore(String time, String description) {
            this.time = time;
            this.description = description;
        }

    }

    protected static final Map<String, MissionConfigurationBo.MissionTimeStore> MISSION_DEFAULT_CONFIG_STORE = new HashMap<>();

    static {
        MISSION_DEFAULT_CONFIG_STORE.put(MISSION_TIME_EXPLORE_KEY, new MissionConfigurationBo.MissionTimeStore(DEFAULT_TIME_EXPLORE, "Base explore time"));
        MISSION_DEFAULT_CONFIG_STORE.put(MISSION_TIME_GATHER_KEY, new MissionConfigurationBo.MissionTimeStore(DEFAULT_TIME_GATHER, "Base gather time"));
        MISSION_DEFAULT_CONFIG_STORE.put(MISSION_TIME_ESTABLISH_BASE_KEY,
                new MissionConfigurationBo.MissionTimeStore(DEFAULT_TIME_ESTABLISH_BASE, "Base establish base time"));
        MISSION_DEFAULT_CONFIG_STORE.put(MISSION_TIME_ATTACK_KEY, new MissionConfigurationBo.MissionTimeStore(DEFAULT_TIME_ATTACK, "Base attack time"));
        MISSION_DEFAULT_CONFIG_STORE.put(MISSION_TIME_CONQUEST_KEY,
                new MissionConfigurationBo.MissionTimeStore(DEFAULT_TIME_CONQUEST, "Base conquest time"));
        MISSION_DEFAULT_CONFIG_STORE.put(MISSION_TIME_COUNTERATTACK_KEY,
                new MissionConfigurationBo.MissionTimeStore(DEFAULT_TIME_COUNTERATTACK, "Base counterattack time"));
        MISSION_DEFAULT_CONFIG_STORE.put(MISSION_TIME_DEPLOY_KEY, new MissionConfigurationBo.MissionTimeStore(DEFAULT_TIME_DEPLOY, "Base deploy time"));
    }

    @PostConstruct
    void init() {
        insertMissionBaseTimeIfMissing();
    }

    public Long findMissionBaseTimeByType(MissionType type) {
        Long retVal;
        switch (type) {
            case EXPLORE:
                retVal = findMissionBaseTime(MISSION_TIME_EXPLORE_KEY);
                break;
            case GATHER:
                retVal = findMissionBaseTime(MISSION_TIME_GATHER_KEY);
                break;
            case ESTABLISH_BASE:
                retVal = findMissionBaseTime(MISSION_TIME_ESTABLISH_BASE_KEY);
                break;
            case ATTACK:
                retVal = findMissionBaseTime(MISSION_TIME_ATTACK_KEY);
                break;
            case COUNTERATTACK:
                retVal = findMissionBaseTime(MISSION_TIME_COUNTERATTACK_KEY);
                break;
            case CONQUEST:
                retVal = findMissionBaseTime(MISSION_TIME_CONQUEST_KEY);
                break;
            case DEPLOY:
                retVal = findMissionBaseTime(MISSION_TIME_DEPLOY_KEY);
                break;
            default:
                throw new SgtBackendInvalidInputException("Unsupported mission base time type, specified: " + type.name());
        }
        return retVal;
    }

    private List<Configuration> findAllMissionBaseTime() {
        return configurationRepository.findByNameIn(createMissionsKeyArray());
    }

    private Long findMissionBaseTime(String key, String defaultValue) {
        return Long.valueOf(configurationBo.findOrSetDefault(key, defaultValue).getValue());
    }

    /**
     * Finds the mission base time, if not set will return the default
     *
     * @param key The mission KEY, should use one of the constants available here
     * @throws ProgrammingException When key is not a vlid mission key
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com
     * @since 0.8.1
     */
    private Long findMissionBaseTime(String key) {
        return findMissionBaseTime(key, MISSION_DEFAULT_CONFIG_STORE.get(key).time);
    }

    private void insertMissionBaseTimeIfMissing() {
        List<Configuration> storedValues = findAllMissionBaseTime();
        MISSION_DEFAULT_CONFIG_STORE.forEach((key, value) -> {
            if (storedValues.stream().noneMatch(current -> current.getName().equals(key))) {
                log.info("Mission time for " + key + " doesn't exists, adding default");
                doSaveDefaultMissionBaseTime(key);
            }
        });

    }

    private ArrayList<String> createMissionsKeyArray() {
        ArrayList<String> missionKeys = new ArrayList<>();
        missionKeys.add(MISSION_TIME_EXPLORE_KEY);
        missionKeys.add(MISSION_TIME_GATHER_KEY);
        missionKeys.add(MISSION_TIME_ESTABLISH_BASE_KEY);
        missionKeys.add(MISSION_TIME_ATTACK_KEY);
        missionKeys.add(MISSION_TIME_CONQUEST_KEY);
        missionKeys.add(MISSION_TIME_COUNTERATTACK_KEY);
        missionKeys.add(MISSION_TIME_DEPLOY_KEY);
        return missionKeys;
    }

    private void doSaveDefaultMissionBaseTime(String key) {
        MissionTimeStore defaultValue = MISSION_DEFAULT_CONFIG_STORE.get(key);
        configurationBo.save(new Configuration(key, defaultValue.time, defaultValue.description));
    }
}

File: rust-backend/owge-business/src/bo/mission_configuration_bo.rs
//! Port of `business/mission/MissionConfigurationBo.java`.
//!
//! Holds the per-mission-type base time (in seconds) in the `configuration`
//! key/value table. The Java bean inserts every default at boot (`@PostConstruct
//! insertMissionBaseTimeIfMissing`); here we read lazily via
//! [`ConfigurationBo::find_or_set_default`], which inserts the default the first
//! time the key is read — same observable result without a boot hook.

use crate::bo::ConfigurationBo;
use crate::db::Db;
use crate::error::{OwgeError, OwgeResult};
use crate::model::mission::MissionType;

/// Configuration keys (mirrors the `MISSION_TIME_*_KEY` constants).
pub const MISSION_TIME_EXPLORE_KEY: &str = "MISSION_TIME_EXPLORE";
pub const MISSION_TIME_GATHER_KEY: &str = "MISSION_TIME_GATHER";
pub const MISSION_TIME_ESTABLISH_BASE_KEY: &str = "MISSION_TIME_ESTABLISH_BASE";
pub const MISSION_TIME_ATTACK_KEY: &str = "MISSION_TIME_ATTACK";
pub const MISSION_TIME_CONQUEST_KEY: &str = "MISSION_TIME_CONQUEST";
pub const MISSION_TIME_COUNTERATTACK_KEY: &str = "MISSION_TIME_COUNTERATTACK";
pub const MISSION_TIME_DEPLOY_KEY: &str = "MISSION_TIME_DEPLOY";

/// Defaults (mirrors the `DEFAULT_TIME_*` constants).
const DEFAULT_TIME_EXPLORE: &str = "60";
const DEFAULT_TIME_GATHER: &str = "900";
const DEFAULT_TIME_ESTABLISH_BASE: &str = "43200";
const DEFAULT_TIME_ATTACK: &str = "600";
const DEFAULT_TIME_CONQUEST: &str = "86400";
const DEFAULT_TIME_COUNTERATTACK: &str = "60";
const DEFAULT_TIME_DEPLOY: &str = "60";

pub struct MissionConfigurationBo;

impl MissionConfigurationBo {
    /// `findMissionBaseTimeByType` — the per-type config key + its default. Reads
    /// (and lazily inserts) the value via `ConfigurationBo::find_or_set_default`.
    /// Unsupported mission types raise `InvalidInput` (the Java
    /// `SgtBackendInvalidInputException`).
    pub async fn find_mission_base_time(db: &Db, mission_type: MissionType) -> OwgeResult<i64> {
        let (key, default) = match mission_type {
            MissionType::Explore => (MISSION_TIME_EXPLORE_KEY, DEFAULT_TIME_EXPLORE),
            MissionType::Gather => (MISSION_TIME_GATHER_KEY, DEFAULT_TIME_GATHER),
            MissionType::EstablishBase => {
                (MISSION_TIME_ESTABLISH_BASE_KEY, DEFAULT_TIME_ESTABLISH_BASE)
            }
            MissionType::Attack => (MISSION_TIME_ATTACK_KEY, DEFAULT_TIME_ATTACK),
            MissionType::Counterattack => {
                (MISSION_TIME_COUNTERATTACK_KEY, DEFAULT_TIME_COUNTERATTACK)
            }
            MissionType::Conquest => (MISSION_TIME_CONQUEST_KEY, DEFAULT_TIME_CONQUEST),
            MissionType::Deploy => (MISSION_TIME_DEPLOY_KEY, DEFAULT_TIME_DEPLOY),
            other => {
                return Err(OwgeError::InvalidInput(format!(
                    "Unsupported mission base time type, specified: {}",
                    other.code()
                )))
            }
        };
        let value = ConfigurationBo::find_or_set_default(db, key, default)
            .await?
            .value;
        value.trim().parse::<i64>().map_err(|_| {
            OwgeError::Common(format!(
                "Mission base time config {key} has a non-numeric value: {value}"
            ))
        })
    }
}
